### PR TITLE
Add scheduled jobs

### PR DESCRIPTION
**Pull request by dcramer**
**WIP**

Basic wire-up of scheduled jobs.

Refs GH-2730

---

This change is [Reviewable](https://reviewable.io/reviews/getsentry/sentry/2743)

### CODE REVIEW

**Comment by dcramer** [comment]
@tkaemming @mattrobenolt thoughts?

File: src/sentry/tasks/scheduler.py
Comment (mattrobenolt) [inline src/sentry/tasks/scheduler.py]: If we had a ton of scheduled jobs, this would be possible that it'd start backing up, since it's only consuming 100 at a time, then waiting 1 minute between.

In practice, I don't think this is an issue since it'll be pretty lightly utilized.

And to be pedantic, this isn't a queryset. ;)

Comment (dcramer) [inline src/sentry/tasks/scheduler.py]: I explicitly chose to limit it to make sure we dont have a similar issue we have in the resolution cleanup (that its unbounded). If it gets backlogged that's fine. We could obviously be more smart and have it requeue itself, but I dont foresee this being an issue.

File: src/sentry/tasks/scheduler.py
Comment (mattrobenolt) [inline src/sentry/tasks/scheduler.py]: Instead of jobs.payload, I'd say we should support both kwargs and args, if the idea is that this is generic for any queued job. Or payload can just contain `{"kwargs": {}, "args": []}`, and we do:  `app.send_task(job.name, **job.payload)` here.

Comment (dcramer) [inline src/sentry/tasks/scheduler.py]: fuck args, that causes too many issues in life. Things can be required to use kwargs. It's safer.

Comment (mattrobenolt) [inline src/sentry/tasks/scheduler.py]: I hope we need none of these: http://docs.celeryproject.org/en/latest/reference/celery.app.task.html#celery.app.task.Task.apply_async

;)

retry and whatnot are useful, but it's probably not that big of a deal.

Comment (dcramer) [inline src/sentry/tasks/scheduler.py]: if we did id rather they be columns to avoid excess db bloat

Comment (mattrobenolt) [inline src/sentry/tasks/scheduler.py]: Yeah, that's fair. In that case, might make sense to name this column `kwargs` instead of `payload`. But whichever works.

**Comment by mattrobenolt** [comment]
Overall, this is what I had in my head (you just beat me to it). +1

**Comment by tkaemming** [comment]
Seems reasonable — I don't have any background on what this is for, though.

**Comment by dcramer** [comment]
Going to merge once green as its needed in getsentry

**Comment by mattrobenolt** [comment]
I'm going to take this over.

File: src/sentry/tasks/scheduler.py
Comment (mattrobenolt) [inline src/sentry/tasks/scheduler.py]: Nah. Just use:

```python
from sentry.app import locks
from sentry.utils import TimedRetryPolicy
...

lock = locks.get('scheduler:process', duration=60)
with TimedRetryPolicy(5)(lock.acquire):
  ...
```

You don't wanna do all the stuff manually.

Comment (tkaemming) [inline src/sentry/tasks/scheduler.py]: What @mattrobenolt said.

If you want to maintain the `nowait=True` behavior from the previous change, you don't have to wrap the `lock.acquire` in the retry policy and can instead do this:

```python
with locks.get('scheduler:process', duration=60).acquire():
```

Comment (ehfeng) [inline src/sentry/tasks/scheduler.py]: Ok fixed. I'm sticking with Matt's behavior, as it looks like it's relatively common in the codebase and some of the scheduled tasks include sending email, so I assume the potential for 3rd party glitches I'd want to retry.

File: src/sentry/tasks/scheduler.py
Comment (tkaemming) [inline src/sentry/tasks/scheduler.py]: Extra newline

File: src/sentry/tasks/scheduler.py
Comment (tkaemming) [inline src/sentry/tasks/scheduler.py]: Not a `queryset`, like @mattrobenolt said.

File: src/sentry/tasks/scheduler.py
Comment (tkaemming) [inline src/sentry/tasks/scheduler.py]: This probably shouldn't be `info`, I'll defer to @jtcunning's judgement here if he's got opinions.

File: src/sentry/tasks/scheduler.py
Comment (tkaemming) [inline src/sentry/tasks/scheduler.py]: Might be worth logging something if the size of this is `> 100` since that could get out of control.

**Review by tkaemming — COMMENTED** [review]
> I'm sticking with Matt's behavior, as it looks like it's relatively common in the codebase and some of the scheduled tasks include sending email, so I assume the potential for 3rd party glitches I'd want to retry.

I'm not sure I understand your rationale here, since this is going to retry every 60 seconds anyway but whatever.

File: src/sentry/tasks/scheduler.py
Comment (JTCunning) [inline src/sentry/tasks/scheduler.py]: `logger = logging.getLogger('sentry.scheduler')`

File: src/sentry/tasks/scheduler.py
Comment (JTCunning) [inline src/sentry/tasks/scheduler.py]: This is a `debug` statement.

File: src/sentry/tasks/scheduler.py
Comment (JTCunning) [inline src/sentry/tasks/scheduler.py]: We should probably be deleting objects one by one after a successful enqueuing. If for some reason we error out on `send_task`, the next time this runs we will reschedule all jobs that were enqueued before the error.

**Review by JTCunning — DISMISSED** [review]

File: src/sentry/tasks/scheduler.py
Comment (ehfeng) [inline src/sentry/tasks/scheduler.py]: I couldn't find @mattrobenolt's comment? Just cmd-F'd over this page...

File: src/sentry/tasks/scheduler.py
Comment (tkaemming) [inline src/sentry/tasks/scheduler.py]: 👌

File: src/sentry/tasks/scheduler.py
Comment (ehfeng) [inline src/sentry/tasks/scheduler.py]: Fixed

File: src/sentry/tasks/scheduler.py
Comment (ehfeng) [inline src/sentry/tasks/scheduler.py]: nvm, going with Ted's behavior.

File: src/sentry/tasks/scheduler.py
Comment (ehfeng) [inline src/sentry/tasks/scheduler.py]: I tried...something? Not sure if it's what you wanted fixed though.

File: src/sentry/tasks/scheduler.py
Comment (ehfeng) [inline src/sentry/tasks/scheduler.py]: Added.

File: src/sentry/tasks/scheduler.py
Comment (ehfeng) [inline src/sentry/tasks/scheduler.py]: Fixed.

File: src/sentry/tasks/scheduler.py
Comment (tkaemming) [inline src/sentry/tasks/scheduler.py]: Not sure where it went (victim of force push at some point maybe) but the gist of the comment is that the variable name doesn't actually represent what the value is, since casting it to `list` causes it not to be a `QuerySet`.

File: src/sentry/tasks/scheduler.py
Comment (ehfeng) [inline src/sentry/tasks/scheduler.py]: Fixed

File: src/sentry/tasks/scheduler.py
Comment (ehfeng) [inline src/sentry/tasks/scheduler.py]: Fixed.

File: src/sentry/tasks/scheduler.py
Comment (ehfeng) [inline src/sentry/tasks/scheduler.py]: Ah, that's what I thought. I moved somethings around (as part of the warning when there's >100 jobs), so `queryset` should be appropriate now.

**Review by JTCunning — COMMENTED** [review]
Logging stuff is good. Deferring rest of this to someone else.

File: src/sentry/tasks/scheduler.py
Comment (tkaemming) [inline src/sentry/tasks/scheduler.py]: Probably better to just select 101 and see if size of the size of the result exceeds 100 to avoid making the database do the query twice. I'm not sure the exact number matters, we just want to be able to figure out if we're lagging behind or not.

File: src/sentry/tasks/scheduler.py
Comment (tkaemming) [inline src/sentry/tasks/scheduler.py]: This should use parameterized logging instead of string formatting like the log statement below.

**Review by tkaemming — APPROVED** [review]
More nitpicky stuff but not going to block on it

File: src/sentry/tasks/scheduler.py
Comment (ehfeng) [inline src/sentry/tasks/scheduler.py]: Good idea. Done.

File: src/sentry/tasks/scheduler.py
Comment (ehfeng) [inline src/sentry/tasks/scheduler.py]: Ah, I did not realized it worked that that. Done.

File: src/sentry/tasks/scheduler.py
Comment (tkaemming) [inline src/sentry/tasks/scheduler.py]: I forget exactly how the query cache works but it'd be safer to at this point just coerce this result to a `list` so that way it's explicit that you don't intend for it to run twice.

File: src/sentry/tasks/scheduler.py
Comment (tkaemming) [inline src/sentry/tasks/scheduler.py]: This will always be 101 if it logs now.

Comment (tkaemming) [inline src/sentry/tasks/scheduler.py]: Also `ScheduledJob's` is a possessive, which this is not.

**Comment by getsentry-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR includes migrations</td>
    </tr>
  </tbody>
</table>

## Migration Checklist

- [ ] new columns need to be nullable (unless table is new)
- [ ] migration with any new index needs to be done concurrently
- [ ] data migrations should not be done inside a transaction
- [ ] before merging, check to make sure there aren't conflicting migration ids


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

**Pull request merged**
